plugins/cassandra/cassandraembed.cpp

```diff
@@ -311,20 +318,20 @@ class CassandraRetryingFuture : public CInterface
             limiter->wait();
         future = cass_session_execute(session, statement);
         if (limiter)
-            cass_future_set_callback(future, signaller, this); // Note - this will call the callback if the future has already completed
+            cass_future_set_callback(future, signaller, LINK(limiter)); // Note - this will call the callback if the future has already completed
     }
     static void signaller(CassFuture *future, void *data)
     {
-        CassandraRetryingFuture *self = (CassandraRetryingFuture *) data;
-        if (self && self->limiter)
-            self->limiter->signal();
+        LinkedSemaphore *sem = (LinkedSemaphore *) data;
+        sem->signal();
```

Comment:
We could assert that it is not NULL - but then again we could do so before any and every pointer dereference in the entire codebase.

Yes, there was a check before. No it wasn't really needed. And bear in mind that despite the prior check, the code before still managed to corrupt memory.